### ISSUE TITLE
[iOS] Native Facebook App Authentication

### ISSUE THREAD

**Issue by lucasgismondi** [facebook/react-native-fbsdk#734]
I'm seeing from others that it is possible to login through the Facebook app for iOS but I can't seem to get it to work. 
I've tried adding 
```- (BOOL)application:(UIApplication *)application openURL:(NSURL *)URL options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options
{
  if ([[FBSDKApplicationDelegate sharedInstance] application:application openURL:URL options:options]) {
    return YES;
  }
  return NO;
}```

but the webview shows when pressing Log In. There's also no way for me to check if this works because I'm not able to sign into a different Facebook account after I've already logged in. Has anyone successfully been able to do this? Am I missing something? Is it even possible? Is there something I need to add in my info.plist?

react-native-fbsdk: v1.1.1
react-native: v0.61.4

Thanks!

**Comment by eroy4u** [comment]
Encounter the same issue. Cannot use Facebook app to log in.

**Comment by Sherif-Magdy** [comment]
+1

**Comment by Jani-10thmuse** [comment]
we have same issue

**Comment by mralj** [comment]
We have same issue, everything works fine, but if you have installed FB app on your phone (which I guess most of the people do), _FB Login Flow_ is broken, meaning

1. I tap on "Login with FB"
2. Modal which displays FB's Web page is opened, there is button "Login With the Facebookapp"
3. I tap on that, I am taken to FB native app, there is button "Continue"
4. I tap on that, I am taken back to point 2. (ie. modal with Web page)

And that basically it, like infinite loop which I can only "exit" by exiting whole "flow", thus not being able to _Login with FB_

**Comment by test-js** [comment]
@mralj Did you do anything specific to get the "Login With the Facebook app" to show in the web page that pops up??

**Comment by mralj** [comment]
Nope, we are not doing  anything "on the native side", and here is our code:

```
  LoginManager.logInWithPermissions(["email"]).then(
    result => {
      const userCanceledFbLoginProcess = !!result && result.isCancelled;
      if (userCanceledFbLoginProcess) {
        return;
      }
      AccessToken.getCurrentAccessToken().then(
        fbResponse => {
          if (!!fbResponse) {
            updateUserProfileWithDataFromFB(fbResponse);
          }
        },
        tokenError => {
          if (!!tokenError) {
            setImmediateErrorAlert("common.fbErrors.cantLoginWithFacebook");
            logError({ error: tokenError });
          }
        }
      );
    },
    error => {
      setImmediateErrorAlert("common.fbErrors.initialProcessingError");
      logError({ error });
    }
  );
```

- I have confirmed this scenario on multiple iOS devices (which have installed FB app). 
- We have latest RN version (0.62.x)
- not sure if this could be causing problems, but we are using [Wix RNN](https://github.com/wix/react-native-navigation)
- Everything worked in the past for sure, but I cannot pinpoint exact "time" problem started occurring

**Comment by mralj** [comment]
Me again 🙂

I have realized that I have proably  misunderstood OP and that we have actually different issues, mine being not "infinite  loop" when FB app installed, and OP's being that FB  app,event hough installed is not opened during Login process (and in my scenario FB  app is being opened and that caused trouble).

Since the FB app is opening for me, and I have fixed issue I was having, I'm sharing our "setup".

JS code is already posted above, so here are 2 other relevatnt files:

### Info.plst 

| Key                                   | Type     | Value           |
|---------------------------------------|----------|-----------------|
| FacebookAdvertiserIDCollectionEnabled | Booelan  | Yes             |
| FacebookAppID                         | String   | YOUR_FB_APP_ID  |
| FacebookAutoLogAppEventsEnabled       | Boolean  | Yes             |
| FacebookDisplayName                   | String   | $(PRODUCT_NAME) |


### AppDelegate.m

In `application didFinishLaunchingWithOptions` I have 

```
[[FBSDKApplicationDelegate sharedInstance] application:application
                           didFinishLaunchingWithOptions:launchOptions];
```

And adding this fixed issue for me (I think this was needed due to th fact I am using RNN)

```
- (BOOL)application:(UIApplication *)app
            openURL:(NSURL *)url
            options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{
  if ([[FBSDKApplicationDelegate sharedInstance] application:app openURL:url options:options]) {
    return YES;
  }
  
  return NO;
}
```

Regarding this: 

> There's also no way for me to check if this works because I'm not able to sign into a different Facebook account after I've already logged in.

Maybe stupid question, but I am not sure why can't you simply logout (and then sign into another account ?

When I was testing this, I was loging-in via _Tests users_. If you were unaware of this feature, here are steps to create them:

1. You go to: https://developers.facebook.com
2. From "My apps" drop-down select your app
3. Go to Roles > Test users
4. Add some test users :) (by clicking edit you can change stuff like password for that test user etc.)

When you are "done" with that account you can simply delete it and choose another one.

If you were testing with your personal FB account, here are my suggestions on "logging out":

1. Go to settings > Apps and Websites > Find your app > Select it > Tap on Remove
2. (this actually logs you out)  Settings > Security And Login > Find device(s) from which you want to log out, tap "menu icon" > Log out

Hope this helps 😄

**Comment by AlexaJosse** [comment]
The solution above worked for me ! The facebook SDK documentation shoul be updated

**Comment by ZChenzy** [comment]
> We have same issue, everything works fine unless you have installed Facebook app on your phone (which I guess most of the people do), _FB Login Flow_ is broken, meaning
> 
> 1. I tap on "Login with FB"
> 2. Modal which displays FB's Web page is opened, there is button "Login With the Facebookapp"
> 3. I tap on that, I am taken to FB native app, there is button "Continue"
> 4. I tap on that, I am taken back to point 2. (ie. modal with Web page)
> 
> And that basically it, like infinite loop which I can only "exit" by exiting whole "flow", thus not being able to _Login with FB_
> 
> I have tested this on both Simulator and device where Facebook app was NOT installed, then login process was done via "web" and it worked as expected.

Did you ever resolve this issue? I am running into this issue

**Comment by mralj** [comment]
Yes, as stated above:

> And adding this fixed issue for me (I think this was needed due to the fact I am using RNN)
> 
> ```
> - (BOOL)application:(UIApplication *)app
>             openURL:(NSURL *)url
>             options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
> {
>   if ([[FBSDKApplicationDelegate sharedInstance] application:app openURL:url options:options]) {
>     return YES;
>   }
>   
>   return NO;
> }
> ```

**Comment by creativemind1** [comment]
> Yes, as stated above:
>
> > And adding this fixed issue for me (I think this was needed due to the fact I am using RNN)
> > ```
> > - (BOOL)application:(UIApplication *)app
> >             openURL:(NSURL *)url
> >             options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
> > {
> >   if ([[FBSDKApplicationDelegate sharedInstance] application:app openURL:url options:options]) {
> >     return YES;
> >   }
> >  
> >   return NO;
> > }
> > ```

My scenario is:
- Safari web facebook should not be logged in
- Facebook App should be logged in
- Now Click on facebook login button from your app, it will take you first to the safari web and then popup comes with Open this page with facebook. You will click on "Open". It will take to facebook app where you will login.
- It will redirect to your App
- Now here I'm getting the error like com.facebook.sdk.core error 3 :(
I just found this bug. Not sure till when it is there. I'm using all the latest FBSDK. Can you help me @mralj

**Comment by mralj** [comment]
I am by no means expert in this, but, according to these ([link1](https://github.com/facebookarchive/facebook-swift-sdk/issues/298#issuecomment-446363978) & [link2](https://stackoverflow.com/questions/53733275/facebook-login-error-unknown-error-building-url-com-facebook-sdk-core-error-3)), this error was known to happen on FB SDK side, we are running `v1.1.2` (of this plugin) and everything seems t be working fine. 

Could you try downgrading and check if issue is still there ?

**Comment by mralj** [comment]
I have updated `react-native-fbsdk` in our project to latest version (`v2.0.0`), and for me everything works fine.

I have tested on 2 devices (and 2 different scenarios), both devices running iOS 13, both builds were `release` :

**1st** scenario was:
- Logged out of Safari, logged-in in FB app
- Was previously logged-in with that FB account in our app

**2nd** scenario was:
- Logged out of Safari, logged-in in FB app
- Created new test account, it was never associated with our app before

So good news is, everything works as expected 🙂
Bad news is I have no idea why SDK doesn't work for you 🤷‍♂️😕

**P.S.**

Obviously this has nothing to do with this particular issue, but also tested on Android "while I was at it", also works :)

**Comment by creativemind1** [comment]
It's good to see it's working for you. If I'm not bugging you much, would you please care to look at my `AppDelegate.m` file.

`// Copyright 2015-present 650 Industries. All rights reserved.

#import "AppDelegate.h"
#import <Firebase.h>
#import <FBSDKCoreKit/FBSDKCoreKit.h>
#import <React/RCTLinkingManager.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  [FIRApp configure];
  return [super application:application didFinishLaunchingWithOptions:launchOptions];
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
  [FBSDKAppEvents activateApp];
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url  sourceApplication:(NSString *)sourceApplication
         annotation:(id)annotation {
  return [[FBSDKApplicationDelegate sharedInstance] application:application
                                                         openURL:url
                                               sourceApplication:sourceApplication
                                                      annotation:annotation];
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
  [super applicationWillEnterForeground:application];
}

#pragma mark - Background Fetch

- (void)application:(UIApplication *)application performFetchWithCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler
{
  [super application:application performFetchWithCompletionHandler:completionHandler];
}

#pragma mark - Handling URLs

- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{

if ([[FBSDKApplicationDelegate sharedInstance] application:application openURL:url options:options]) {
    return YES;
  }
   

  if ([RCTLinkingManager application:application openURL:url options:options]) {
    return YES;
  }

  return NO;
}

- (BOOL)application:(UIApplication *)application continueUserActivity:(NSUserActivity *)userActivity restorationHandler:(void (^)(NSArray<id<UIUserActivityRestoring>> * _Nullable))restorationHandler
{
  return [super application:application continueUserActivity:userActivity restorationHandler:restorationHandler];
}

#pragma mark - Notifications

- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)token
{
  [super application:application didRegisterForRemoteNotificationsWithDeviceToken:token];
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)err
{
  [super application:application didFailToRegisterForRemoteNotificationsWithError:err];
}

- (void)application:(UIApplication *)application didReceiveRemoteNotification:(NSDictionary *)userInfo fetchCompletionHandler:(void (^)(UIBackgroundFetchResult result))completionHandler
{
  [super application:application didReceiveRemoteNotification:userInfo fetchCompletionHandler:completionHandler];
}

@end

`

**Comment by mralj** [comment]
TBH I have no idea, your `AppDelegate` file is. much more complicated than ours, and I have no experience in iOS development / Objective-C.

This is wild shot in the dark, but it _looks_ to me that you have polymorphic `openURL`

### 1

```
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url  sourceApplication:(NSString *)sourceApplication
         annotation:(id)annotation {
  return [[FBSDKApplicationDelegate sharedInstance] application:application
                                                         openURL:url
                                               sourceApplication:sourceApplication
                                                      annotation:annotation];
}
```

and

### 2
```
- (BOOL)application:(UIApplication *)application
   openURL:(NSURL *)url
   options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options
{

if ([[FBSDKApplicationDelegate sharedInstance] application:application openURL:url options:options]) {
    return YES;
  }
   

  if ([RCTLinkingManager application:application openURL:url options:options]) {
    return YES;
  }

  return NO;
}
```

Looking at iOS source code I found this:

```
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(nullable NSString *)sourceApplication annotation:(id)annotation API_DEPRECATED_WITH_REPLACEMENT("application:openURL:options:", ios(4.2, 9.0)) API_UNAVAILABLE(tvos);


typedef NSString * UIApplicationOpenURLOptionsKey NS_TYPED_ENUM;

- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options API_AVAILABLE(ios(9.0)); // no equiv. notification. return NO if the application can't open for some reason
```

Could it be that first one is causing problem ?
**if** I am reading code correctly, isn't  **1st** method deprecated (on iOS 9+) ?

**Comment by creativemind1** [comment]
yes it is deprecated. In fact, it is showing in my Xcode as well. But I still don't understand why the code is fetching from it. This is directly using the FBSDK framework from the node_modules/react-native-fbsdk.

**Comment by mralj** [comment]
Maybe stupid question, but can’t you simply delete it and see if everything works ?

The `openUrl` scenario should be covered by **2nd** “case”

**Comment by creativemind1** [comment]
Ok I will give this a try. Just a last question if you don't mind,
I have not added swift dependency from this URL https://github.com/facebook/facebook-ios-sdk. as shown in the facebo?ok Get Started Link. Is it compulsory

**Comment by mralj** [comment]
I don't think it is necessary to addd that dependency, IMHO, you should only [follow steps](https://github.com/facebook/react-native-fbsdk#installation) from this library. For us everything worked fine by just following those steps. 
This way `react-native-fbsdk` will take care of adding all deps needed.

Here is example from our `Podfile.lock`:

```
- FBSDKCoreKit (7.0.0):
    - FBSDKCoreKit/Basics (= 7.0.0)
    - FBSDKCoreKit/Core (= 7.0.0)
  - FBSDKCoreKit/Basics (7.0.0)
  - FBSDKCoreKit/Core (7.0.0):
    - FBSDKCoreKit/Basics
  - FBSDKLoginKit (7.0.0):
    - FBSDKLoginKit/Login (= 7.0.0)
  - FBSDKLoginKit/Login (7.0.0):
    - FBSDKCoreKit (~> 7.0.0)
  - FBSDKShareKit (7.0.0):
    - FBSDKShareKit/Share (= 7.0.0)
  - FBSDKShareKit/Share (7.0.0):
    - FBSDKCoreKit (~> 7.0.0)**
```

As you can see we are not using "Swift version"

**Comment by creativemind1** [comment]
My bad. The issue was there in my `Info.plist` file. I had to change the boolean value of `FacebookAutoInitEnabled` to `true` and it is now working fine. Now even my Facebook Pixel events are working. So these are my changes:
```
<key>FacebookAdvertiserIDCollectionEnabled</key>
<true/>
<key>FacebookAutoLogAppEventsEnabled</key>
<true/>
<key>FacebookAutoInitEnabled</key>
<true/>

```

Your help was very appreciated, thanks. @mralj
